### DISCUSSION TITLE
leftwm NetBSD package on the way

### DISCUSSION THREAD

**Discussion by 0323pin** [leftwm/leftwm#590] [General]
Hi, pkgsrc maintainer here.

After a bit of a bumpy ride, will come to that soon, I've been running `leftwm` on my development machine for almost a day.
I'm enjoying this wm and would like to merge it into the pkgsrc main branch, perhaps when 0.2.10 is official released, as that was what I have packaged and am running, built from git-HEAD through pkgsrc. Now to the bumpy ride 😃 ...

The README.md states the follwoing

> List of LeftWM dependencies:
xorg (libxinerama, xrandr, xorg-server)
bash
rust

-Rust is an obvious one, though only required for building.
-bash ?!? What am I missing, why is this required? I don't even have `bash` installed and the package builds and runs fine without it. Will I be hitting my head in the next couple of days? Personally, I run `nushell` but, I have the 3 base shells installed, `sh`, `csh` and `ksh`. I would prefer to keep `bash` out of the package and so far, I see no reason to include it as dependency.
-xorg, here I think you could be slightly more specific and this is where the bumpy ride started.

Ok, `xorg-server` and `xrandr` are included in NetBSD base, so I installed `libxinerama`, built and installed the wm, picked a theme and a `config.toml` from github and run `leftwm-check` and all was good. Logout and login and it would take no input whatsoever. The only way to get out was to press and hold _power_. 

Fortunately, I was able to see the error messages about a thread-panic, so I've repeated the procedure a few times while trying to catch why there was a thread-panic and finally saw missing `libx11.so`.

So, I've linked against `libx11` at compile time and now, all is working fine. Even posted it on [r/unixporn](https://www.reddit.com/r/unixporn/comments/r5ko33/leftwm_just_one_fix/) after some _ricing_.

Perhaps save some people the headache and declare `libx11` as dependency, It would have saved me sometime.

Finally, one question. If I bind the terminal to `alacritty` it launches immediately but, if I bind it to `xterm`, it takes nearly 10 seconds to display. Do note that other application, like my web browser, launch normally, the delay is only with `xterm`. Is this a known problem?

I'd like to keep using `xterm` if possible, as it's included in the NetBSD base but, will most probably start using `alacritty` if I don't get it to launch faster.

Anyway, thank you for any comments and looking forward to the 0.2.10 release and to merge `leftwm` into pkgsrc.

Best,
pin

**Comment by mautamu** [comment]
Howdy, and welcome fellow adventurer!

Thank you for taking on this issue and for reporting the problems that you have found! Glad to hear things are moving along for NetBSD!

I apologize that this is out of order:
- Xterm: The `xterm` issue is being resolved by #587, not sure exactly when that will be merged, but it *seems* to be nearly complete to me. (Tagging @AethanFoot)
- libx11: I agree, this could be made more clear. If you wouldn't mind submitting a PR against the README, that would be great. I believe we have stuck with a generic `xorg` as it's the package for Arch-based distributions. If I'm understanding correctly, what it *ought* to read is "`xorg` (specifically, `libx11`, `xrandr`, `xorg-server`, `libxinerama`)"? I *believe* those are the packages I needed to get it working on gKISS Linux not too long ago, but might be forgetting one or two. 
- Bash: Bash is not *required* but many of the themes have decided to use bash rather than pure-posix shell, so it's really an optional dependency. At one point or another I believe the codebase *had* required bash as it used that for launching the `up` and `down` scripts, but now it implicitly calls `sh` so we *ought* to be good to drop the `bash` requirement, and instead label it what it is: an optional requirement for most themes. [Source](https://github.com/leftwm/leftwm/blob/c26b9ef69a9b96ce0bf3a9e2f78c06503eb56c0c/leftwm-core/src/utils/child_process.rs)

Ergo, the new requirements ought to be:
- `cargo` or other `rustc` driver (build): >= 1.52.0
- `xorg` (specifically `libx11`, `xrandr`, `xorg-server`, `libxinerama`) (runtime, build): recent
- `sh`: any posix-compliant shell for starting `up` and `down` files
- `bash`(***optional***): recent

Best,
Mau

**Comment by 0323pin** [comment]
@mautamu thanks for the answer regarding `xterm`, looking forward to check out the fix.

> Glad to hear things are moving along for NetBSD!

No problem, the build itself was a smooth ride :)
The package is nearly ready to be merged, just some minor style and preferences to decide upon.
Anyway, not really in a hurry, as mentioned I'll wait for 0.2.10 as I think a man page could be useful. The separation of themes is rather nice but, it's a bit of a brain twister when you see it for the first time ;)

Surely, I can make a PR tomorrow but, you got it all covered on your reply. I had `libx11` installed but, if I don't link it at compile time the binary won't find it. Most probably because NetBSD doesn't use _"standard paths"_. With that said could be that other things are detected but, not declared in the package. We will see if the build fails on a sandbox. If so, I'll correct it and add what is missing. I don't have enough resources to set-up a bulk-build sandbox myself.

If I have any further questions I will let you know. And, I'll post back once the package is merged.

**Reply by VuiMuich** [reply]
Would we want to have these info in the Readme immediately, or should we wait until the merge of the NetBSD package and include info about it in the same PR as well?

**Reply by 0323pin** [reply]
If you prefer, wait until I push the package. Although, these are the _relevant_ dependencies anyway.

One commit worries me, though https://github.com/leftwm/leftwm/pull/508/commits

If so please introduce a build option, else this means a lot of trouble.

**Reply by AethanFoot** [reply]
I don't think this will be a hard dependency and will only active when systemd logging is available. I would also suppose once we have set that up and better error logging in general, we may be able to expand this to different systems used by other systems. This is mainly to default to logging so the user can debug, it just so happens journald is currently the main thing we use (again it won't be a hard dependency and also users will be able to disable it).
Thanks!

**Reply by 0323pin** [reply]
@AethanFoot Thanks, that's good to know, as long as there is a build option to disable it, I'm fine with it.

**Reply by VuiMuich** [reply]
I will mark #508 draft for now anyways, as I right now have a hard time getting this moving.
But I intended to introduce a cargo feature gate so you can build explicitly without journald support.

**Reply by 0323pin** [reply]
Awesome, thanks. I'll fix the last rough corners on the package tomorrow and have it ready to merge whenever 0.2.10 is out.

Still need to make-up my mind if I should install the default themes or, just leave it to the user to decide. Another option is not to install them by default but, have a build option to add them when building from source.

**Reply by 0323pin** [reply]
I've finished the package now and it's ready to be merged into the main pkgsrc branch, as soon as 0.2.10 is out 😄

**Comment by 0323pin** [comment]
@AethanFoot just confirming #587 fixes the `xterm` issue. Running on the latest head right now 😄

**Comment by 0323pin** [comment]
@mautamu, @VuiMuich, @AethanFoot As promissed, `leftwm` is now on pkgsrc 😄 

https://mail-index.netbsd.org/pkgsrc-changes/2021/12/06/msg244592.html

I'll add install instructions to the README tomorrow.
Thanks!

**Comment by 0323pin** [comment]
Just for information, in case you would like to know. pkgsrc is not only the default package manager for NetBSD and SmartOS but also, supports 18 other platforms.

See, http://www.pkgsrc.org/#index4h1 for a list of supported platforms.

I'm not saying `leftwm` will build without an itch on all of these but, chances are it will build on quite a few of them. We will see in the coming weeks. There are binary builds published regularly for illumos and macOS.

**Comment by 0323pin** [comment]
Following-up on my previous post, it builds on macOS as is 😄 

![sample](https://user-images.githubusercontent.com/90570748/145407920-6e805d56-059a-4975-8b0b-4888e506fe0d.PNG)

**Reply by VuiMuich** [reply]
> Following-up on my previous post, it builds on macOS as is 😄 
> 
> 
> 
> ![sample](https://user-images.githubusercontent.com/90570748/145407920-6e805d56-059a-4975-8b0b-4888e506fe0d.PNG)
> 
> 
> 
> 

Has macOS a deeply buried X11?

**Reply by 0323pin** [reply]
It's possible to set it up, https://www.xquartz.org/ and to use pkgsrc on it, https://pkgsrc.joyent.com/install-on-osx/

I haven't done this myself but, several NetBSD devs use macOS with pkgsrc, so yes, they could in practice run `leftwm` on macOS now.
